linalg/src/wasm/tests.rs

```diff
@@ -0,0 +1,340 @@
+use proptest::prelude::*;
+
+use crate::{
+    frame::mmm::kernel::MatMatMulKer,
+    generic::GenericMmm4x4,
+    mmm::{FusedKerSpec, OutputStoreKer, RoundingPolicy},
+    wasm::WasmMmm4x4,
+};
+
+fn f32_4() -> impl Strategy<Value = [f32; 4]> {
+    [any::<f32>(); 4]
+}
+
+fn f32_4x4() -> impl Strategy<Value = [[f32; 4]; 4]> {
+    [[any::<f32>(); 4]; 4]
+}
+
+fn remove_nan(mut a: [[f32; 4]; 4]) -> [[f32; 4]; 4] {
+    for i in 0..4 {
+        for j in 0..4 {
+            if a[i][j].is_nan() {
+                a[i][j] = 12345678.9;
+            }
+        }
+    }
+    a
+}
+
+fn run(
+    mut initial: [[f32; 4]; 4],
+    elem: FusedKerSpec<f32>,
+) -> ((isize, [[f32; 4]; 4]), (isize, [[f32; 4]; 4])) {
+    let mut output = [[0.0; 4]; 4];
+
+    let spec = [
+        FusedKerSpec::Clear,
+        FusedKerSpec::AddUnicast(OutputStoreKer {
+            ptr: initial.as_mut_ptr() as *mut u8,
+            row_byte_stride: 16,
+            col_byte_stride: 4,
+            item_size: 4,
+        }),
+        elem,
+        FusedKerSpec::Store(OutputStoreKer {
+            ptr: output.as_mut_ptr() as *mut u8,
+            row_byte_stride: 16,
+            col_byte_stride: 4,
+            item_size: 4,
+        }),
+        FusedKerSpec::Done,
+    ];
+
+    let r1 = WasmMmm4x4::kernel(&spec);
+    let actual = remove_nan(output.clone());
+
+    let r2 = GenericMmm4x4::<f32, f32, f32>::kernel(&spec);
+    let expected = remove_nan(output.clone());
+
+    ((r1, actual), (r2, expected))
+}
+
+proptest! {
```

Comment:
what's the logic to add these extra tests ? Are they covering more than the regular test_mmm_kernel_f32 ? I'm not asking for their removal, just trying to understand if there is something missing in the regular standard kernel test suite.

Reply:
Yes, that was exactly the motivation for adding these tests. I wanted to get some confidence that I didn't introduce bugs. I noticed that `test_mmm_kernel_f32` was not covering everything. For example, in leaky relu removing the last line still passes `test_mmm_kernel_f32` (IIRC).

Reply:
Can we make the generic test cases cover these instead ? That would benefit all the kernels, including the future wasm ones that will use a different geometry than 4x4.

Reply:
I'll try to add a test that covers the missing leaky relu case that I know of. I don't know how to find all missing cases though.

Reply:
Sure, the generic kernel test suite is a work in progress, it does not cover corner case and will never do. But it is very beneficial to augment it instead of adding tests that are specialized for a given kernel. Now that you have setup the basic building blocks for wasm kernels, I expect we will pretty soon have a half a dozen or more wasm kernels, so believe me, being able to rely on the generic kernel test suite will make a hug lot of sense...

Reply:
I fixed the existing leaky relu test to multiply by 1 instead of 0 (which was causing the test to be ineffective because multiplying by 0 always gives 0). I also removed the custom Wasm tests.